spi_nand_flash/src/dhara_glue.c

```diff
@@ -3,329 +3,249 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  *
- * SPDX-FileContributor: 2015-2023 Espressif Systems (Shanghai) CO LTD
+ * SPDX-FileContributor: 2015-2024 Espressif Systems (Shanghai) CO LTD
  */
 
 #include <string.h>
+#include <sys/lock.h>
 #include "dhara/nand.h"
+#include "dhara/map.h"
+#include "dhara/error.h"
 #include "esp_check.h"
 #include "esp_err.h"
-#include "freertos/FreeRTOS.h"
-#include "freertos/task.h"
 #include "spi_nand_oper.h"
-#include "spi_nand_flash.h"
+#include "spi_nand_api.h"
 #include "nand.h"
 
-#define ROM_WAIT_THRESHOLD_US 1000
+static struct dhara_nand dhara_nand[NAND_MAX_DEVICES];
+static struct dhara_map dhara_map[NAND_MAX_DEVICES];
+static spi_nand_flash_device_t *s_handle_t[NAND_MAX_DEVICES] = {NULL};
+static uint8_t s_handle_index;
+static _lock_t s_handle_index_lock;
```

Comment:
Can we no longer use the "containerof" trick to get the `spi_nand_flash_device_t` handle from `const struct dhara_nand *n`? As I understand, you need to store the handle index for the duration of dhara call to know which handle the callbacks are related to... Seems like the original "containerof" approach solved that nicely without the extra locks.

Reply:
Oh I see now. That's because you no longer put the dhara_map into spi_nand_flash_device_t. I think you can still do that, via an intermediate holder structure.

This allows you to get rid of all the static members and also NAND_MAX_DEVICES for good measure :)


```c++
struct spi_nand_flash_device_t {
    uint8_t dev_index;
    spi_nand_flash_config_t config;
    spi_nand_chip_t chip;
    const spi_nand_ops *ops;
    void* ops_ctx;   // new member
    uint8_t *work_buffer;
    uint8_t *read_buffer;
    SemaphoreHandle_t mutex;
};

typedef struct {
    struct dhara_nand dhara_info;
    spi_nand_flash_device_t * parent_handle;
} spi_nand_flash_dhara_ctx_t;

// in dhara_init:
static esp_err_t dhara_init(spi_nand_flash_device_t *handle)
{
    // create a holder structure for dhara context
    spi_nand_flash_dhara_ctx_t * dhara_ctx = calloc(1, sizeof(spi_nand_flash_dhara_ctx_t));
    // save the holder inside the device structure
    handle->ops_ctx = dhara_ctx;
    // store the pointer back to device structure in the holder stucture
    dhara_ctx->parent_handle = handle;
    dhara_ctx->dhara_info->log2_page_size = handle->chip.log2_page_size;
    // etc


// in dhara callbacks:
int dhara_nand_is_bad(const struct dhara_nand *n, dhara_block_t b)
{
    // get spi_nand_flash_dhara_ctx_t using containerof
    spi_nand_flash_dhara_ctx_t *dhara_ctx = __containerof(n, spi_nand_flash_dhara_ctx_t, dhara_info);
    // get spi_nand_flash_device_t 
    spi_nand_flash_device_t *dev_handle = dhara_ctx->parent_handle;
    // use dev_handle....

// in dhara_deinit:
// don't forget to delete holder structure (spi_nand_flash_dhara_ctx_t)
```

Reply:
I was aiming to eliminate the dependency on Dhara headers, except in dhara_glue. This approach is the most effective way to achieve this, rather than relying on static variables. Thank you.